File: mobile_frontend/bumpbox_panel/lib/services/detection_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/detection_result.dart';

/// Service for managing ESP32 camera detection
class DetectionService {
  /// Trigger ESP32 camera capture via backend
  ///
  /// Sends a request to the backend which sets a flag that the ESP32
  /// will poll and detect. When detected, ESP32 will capture and send
  /// image for object detection.
  static Future<void> triggerCapture({String? lockerId}) async {
    try {
      final response = await http.post(
        Uri.parse(ApiConfig.triggerCaptureUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'lockerId': lockerId ?? ApiConfig.defaultLockerId}),
      );

      if (response.statusCode != 200) {
        throw Exception(
          'Failed to trigger capture: ${response.statusCode} ${response.body}',
        );
      }

      final data = jsonDecode(response.body);
      if (data['success'] != true) {
        throw Exception('Backend returned failure: ${data['message']}');
      }
    } catch (e) {
      throw Exception('Failed to trigger capture: $e');
    }
  }

  /// Fetch latest detection result from backend
  ///
  /// Returns null if no detection is available or if the detection
  /// is older than the provided 'since' timestamp.
  static Future<DetectionResult?> fetchLatestDetection({
    DateTime? since,
  }) async {
    try {
      final uri = Uri.parse(ApiConfig.latestDetectionUrl).replace(
        queryParameters: since != null
            ? {'since': since.toIso8601String()}
            : null,
      );

      final response = await http.get(uri);

      if (response.statusCode != 200) {
        throw Exception(
          'Failed to fetch detection: ${response.statusCode} ${response.body}',
        );
      }

      final data = jsonDecode(response.body);

      // Backend returns { detection: null } if no detection available
      if (data['detection'] == null) {
        return null;
      }

      return DetectionResult.fromJson(data);
    } catch (e) {
      // For polling, we don't want to throw on network errors
      // Just return null and let the caller retry
      print('[DetectionService] Error fetching detection: $e');
      return null;
    }
  }

  /// Poll for detection result with timeout
  ///
  /// Continuously polls the backend for a detection result until one
  /// is available or the timeout is reached. Returns the detection
  /// result or null if timeout occurs.
  static Future<DetectionResult?> pollForDetection({
    Duration? timeout,
    Duration? pollInterval,
    DateTime? since,
  }) async {
    final timeoutDuration = timeout ?? ApiConfig.detectionTimeout;
    final pollIntervalDuration = pollInterval ?? ApiConfig.pollInterval;
    final startTime = DateTime.now();
    final sinceTime = since ?? startTime;

    while (DateTime.now().difference(startTime) < timeoutDuration) {
      final result = await fetchLatestDetection(since: sinceTime);

      if (result != null) {
        return result;
      }

      // Wait before next poll
      await Future.delayed(pollIntervalDuration);
    }

    // Timeout reached
    return null;
  }
}

File: bumpbox/src/services/detectionService.ts
import axios from "axios";
import { API_ENDPOINTS, apiUrl } from "../api/config";
import {
    DETECTION_POLL_INTERVAL_MS,
    DETECTION_TIMEOUT_MS,
    DEFAULT_LOCKER_ID,
} from "../config/pricingConfig";
import { parseDetectionResult, type DetectionResult } from "../types/detection";

export async function triggerCapture(lockerId = DEFAULT_LOCKER_ID): Promise<void> {
    const response = await axios.post(
        apiUrl(API_ENDPOINTS.triggerCapture),
        { lockerId },
        { headers: { "Content-Type": "application/json" } },
    );

    const data = response.data as Record<string, unknown>;
    if (data.success !== true) {
        throw new Error(`Backend returned failure: ${data.message}`);
    }
}

export async function fetchLatestDetection(since?: Date): Promise<DetectionResult | null> {
    try {
        const params = since ? { since: since.toISOString() } : undefined;
        const response = await axios.get(apiUrl(API_ENDPOINTS.latestDetection), { params });
        const data = response.data as Record<string, unknown>;

        if (data.detection == null) return null;

        return parseDetectionResult(data);
    } catch {
        return null;
    }
}

export async function pollForDetection(options?: {
    timeoutMs?: number;
    pollIntervalMs?: number;
    since?: Date;
}): Promise<DetectionResult | null> {
    const timeoutMs = options?.timeoutMs ?? DETECTION_TIMEOUT_MS;
    const pollIntervalMs = options?.pollIntervalMs ?? DETECTION_POLL_INTERVAL_MS;
    const since = options?.since ?? new Date();
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
        const result = await fetchLatestDetection(since);
        if (result) return result;
        await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }

    return null;
}

export function getLatestImageUrl(): string {
    return `${apiUrl(API_ENDPOINTS.latestImage)}?t=${Date.now()}`;
}
